lib/recursive-methods.rb

```diff
@@ -1,49 +1,198 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
+# https://github.com/stupendousC/recursion-writing
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if (n == 1) || (n == 0)
+    return 1
+  elsif n > 1
+    return n * factorial(n-1)
+  else
+    raise ArgumentError, "argument must be >= 0"
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  # accepts a string and returns the reverse of the string by reversing all letters and all words in the string
+  # ex: reverse("hello world") will convert the input string to "dlrow olleh"
+  if s.length <= 1
+    return s 
+  else
+    return s[-1] + reverse(s[1...-1]) + s[0]
+  end
+  
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(???)
+# Space complexity: O(???)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+  #  accepts a string as a parameter and then reverses the string IN PLACE using a recursive algorithm.
+  left_index = 0
+  right_index = s.length - 1
+  
+  swap(s, left_index, right_index)
+  
+  return s
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# recursive fcn to be used inside reverse_inplace()
+def swap(s, left_index, right_index)
+  if left_index >= right_index
+    return 
+  else
+    temp = s[left_index] 
+    s[left_index] = s[right_index]
+    s[right_index] = temp
+    left_index += 1
+    right_index -= 1
+    swap(s, left_index, right_index)
+  end
+end
+
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  #   N represents a number of bunnies and each bunny has two big floppy ears. We want to compute the total number of ears across all the bunnies recursively (without loops or multiplication).
+  if n == 0
+    # base case: no bunnies left
+    return 0
+  elsif n >= 1
+    # recurse
+    return 2 + bunny(n-1)
+  else
+    # no negative bunnies
+    raise ArgumentError
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def nested(s)
```

Comment:
This is only tangentially recursive, you have a lot of nested loop sand if-else blocks.  

This is an example (inefficient) version with recursion. 

```ruby
  return true if s.empty?
  return false unless s[0] == "(" && s[-1] == ")"
  return nested(s[1..-2])
```